src/sentry/tasks/on_demand_metrics.py

```diff
@@ -0,0 +1,273 @@
+from __future__ import annotations
+
+import logging
+from typing import Any, Dict, List
+
+import sentry_sdk
+from celery.exceptions import SoftTimeLimitExceeded
+
+from sentry import options
+from sentry.api.utils import get_date_range_from_params
+from sentry.models.dashboard_widget import DashboardWidgetQuery, DashboardWidgetTypes
+from sentry.models.organization import Organization
+from sentry.relay.config.metric_extraction import on_demand_metrics_feature_flags
+from sentry.search.events import fields
+from sentry.search.events.builder import QueryBuilder
+from sentry.search.events.types import QueryBuilderConfig
+from sentry.snuba.dataset import Dataset
+from sentry.snuba.referrer import Referrer
+from sentry.tasks.base import instrumented_task
+from sentry.utils import metrics
+from sentry.utils.cache import cache
+from sentry.utils.query import RangeQuerySetWrapper
+
+logger = logging.getLogger("sentry.tasks.on_demand_metrics")
+
+# TTL for cardinality check
+_WIDGET_QUERY_CARDINALITY_TTL = 3600 * 24 * 7  # Cardinality outcome is valid for 7 days.
+
+
+def _get_widget_processing_batch_key():
+    return "on-demand-metrics:widgets:currently-processing-batch"
+
+
+def _get_widget_query_cardinality_cache_key(widget_query: DashboardWidgetQuery):
+    return f"check-widget-query-cardinality:{widget_query.id}"
+
+
+def _set_currently_processing_batch(current_batch):
+    cache.set(_get_widget_processing_batch_key(), current_batch, timeout=3600)
+
+
+def _set_cardinality_cache(cache_key, is_low_cardinality):
+    cache.set(cache_key, is_low_cardinality, timeout=_WIDGET_QUERY_CARDINALITY_TTL)
+
+
+def _get_current_processing_batch(total_batches):
+    previous_batch = cache.get(_get_widget_processing_batch_key(), 0)
+    current_batch = (previous_batch + 1) % total_batches
+    return current_batch
+
+
+def _get_batch_for_widget_query(widget_query_id, total_batches):
+    return widget_query_id % total_batches
+
+
+class HighCardinalityWidgetException(Exception):
+    pass
+
+
+@instrumented_task(
+    name="sentry.tasks.on_demand_metrics.schedule_on_demand_check",
+    queue="on_demand_metrics",
+    max_retries=0,
```

Comment:
I don't think so? I want to avoid re-schedules since it might increase the total number of parallel child tasks running, thereby increasing load on clickhouse.